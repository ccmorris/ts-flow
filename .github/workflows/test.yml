# Run unit tests with Bun on every push

name: CI Tests

on:
  push:
    branches:
      - '*'
  pull_request:
    branches:
      - main

jobs:
  test:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout repository
        uses: actions/checkout@v3

      - name: Set up Bun
        uses: oven-sh/setup-bun@v1
        with:
          bun-version: latest # You can specify a version like '1.0.0' if needed

      - name: Install dependencies
        run: bun install

      - name: Run tests
        run: bun test
