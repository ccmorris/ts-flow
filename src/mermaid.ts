import type {
  ActivityDefinition,
  ChoiceDefinition,
  TaskDefinitions,
  WorkflowResult,
} from './types'

const formatTaskId = (name: string): string => name.replaceAll(' ', '_')

export const toMermaid = (
  tasks: TaskDefinitions,
  result?: WorkflowResult
): string => {
  const failedTaskName =
    result?.success === false && result.transitions.at(-1)?.to?.name
  const tasksWithTraced = tasks.map((task) => {
    const tracedTask = result?.transitions?.find((transition) => {
      return transition.from?.name === task.name
    })
    return {
      ...task,
      isTraced: !!tracedTask,
      tracedTaskName: tracedTask?.transitionName,
      isFailed: task.name === failedTaskName,
    }
  })
  const lines = [
    'flowchart TD',
    `Start((start))${result ? ':::traced==>' : '-->'}`,
    ...tasksWithTraced.map((task) => {
      return task.type === 'activity'
        ? activityToMermaid(task)
        : choiceToMermaid(task)
    }),
  ]
  if (result?.success) {
    // Add the traced style to the End node
    lines[lines.length - 1] = `${lines[lines.length - 1]}:::traced`
  }
  if (result) {
    lines.push('classDef traced fill:green,stroke-width:4px;')
  }
  if (result?.success === false) {
    lines.push('classDef failed fill:#cc0000,stroke-width:4px;')
  }
  return `${lines.join('\n    ')}`
}

const activityToMermaid = (
  activity: ActivityDefinition & {
    isTraced: boolean
    tracedTaskName?: string
    isFailed: boolean
  }
): string => {
  const from = `${formatTaskId(activity.name)}[${activity.name}]`
  const className = activity.isFailed
    ? ':::failed'
    : activity.isTraced
    ? ':::traced'
    : ''
  const isTracedPath = activity.tracedTaskName === 'then'
  const isTracedToEnd =
    activity.then === null && activity.tracedTaskName === '(end)'
  const nextArrow = isTracedPath || isTracedToEnd ? '==>' : '-->'
  const nextDestination = activity.then
    ? formatTaskId(activity.then)
    : 'End((end))'
  const next = `${from}${className}${nextArrow}|then|${nextDestination}`

  const catchConfig = activity.catch
    ? Object.entries(activity.catch).map(([error, catchActivity]) => {
        const from = formatTaskId(activity.name)
        const to = formatTaskId(catchActivity.then ?? 'End((end))')
        const isTracedPath = activity.tracedTaskName === error
        const arrow = isTracedPath ? '==>' : '-->'
        return `\n    ${from}${arrow}|catch ${error}|${to}`
      })
    : ''
  return `${next}${catchConfig}`
}

const choiceToMermaid = (
  choice: ChoiceDefinition & {
    isTraced: boolean
    tracedTaskName?: string
    isFailed: boolean
  }
): string => {
  return `${Object.entries(choice.choices)
    .map(([key, target]) => {
      const from = `${formatTaskId(choice.name)}{${choice.name}}`
      const className = choice.isFailed
        ? ':::failed'
        : choice.isTraced
        ? ':::traced'
        : ''
      const isTracedToEnd = target === null && choice.tracedTaskName === '(end)'
      const isTracedPath = choice.tracedTaskName === key || isTracedToEnd
      const arrow = isTracedPath ? '==>' : '-->'
      const destination = formatTaskId(target ?? 'End((end))')
      return `${from}${className}${arrow}|${key}|${destination}`
    })
    .join('\n    ')}`
}

const encodeMermaidLiveState = (
  mermaidString: string,
  theme?: 'dark' | 'light'
): string => {
  const mermaidLiveState = {
    code: mermaidString,
    mermaid: theme ? JSON.stringify({ theme }) : undefined,
    autoSync: true,
    rough: false,
    updateDiagram: true,
  }
  const base64Encoded = Buffer.from(JSON.stringify(mermaidLiveState)).toString(
    'base64'
  )
  return `base64:${base64Encoded}`
}

export const toMermaidLiveEdit = (
  tasks: TaskDefinitions,
  result?: WorkflowResult
): string => {
  const mermaidString = toMermaid(tasks, result)
  const encodedState = encodeMermaidLiveState(mermaidString, 'dark')
  return `https://mermaid.live/edit#${encodedState}`
}
export const toMermaidPngUrl = (
  tasks: TaskDefinitions,
  result?: WorkflowResult
): string => {
  const mermaidString = toMermaid(tasks, result)
  const encodedState = encodeMermaidLiveState(mermaidString)
  return `https://mermaid.ink/img/${encodedState}?type=png`
}
