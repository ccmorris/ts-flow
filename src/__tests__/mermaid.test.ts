import { describe, expect, test } from 'bun:test'

import type { ActivityDefinitions } from '../types'
import { toMermaid, toMermaidLiveEdit, toMermaidPngUrl } from '../mermaid'

describe('toMermaid', () => {
  test('should generate a mermaid diagram', () => {
    const activities: ActivityDefinitions = [
      {
        name: 'activity1',
        start: true,
        fn: async () => {},
        then: 'activity2',
      },
      {
        name: 'activity2',
        fn: async () => {},
        then: 'activity3',
      },
      {
        name: 'activity3',
        fn: async () => {},
        then: null,
      },
    ]

    const result = toMermaid(activities)

    expect(result).toBe(`flowchart TD
    Start((start))-->
    activity1-->|then|activity2
    activity2-->|then|activity3
    activity3-->End((end))`)
  })

  test('should generate a mermaid diagram with catches', () => {
    const activities: ActivityDefinitions = [
      {
        name: 'activity1',
        start: true,
        fn: async () => {},
        then: 'activity2',
        catch: { timeout: { then: 'timeout_handler' } },
      },
      {
        name: 'activity2',
        fn: async () => {},
        then: 'activity3',
        catch: { timeout: { then: 'timeout_handler' } },
      },
      {
        name: 'activity3',
        fn: async () => {},
        then: null,
      },
      {
        name: 'timeout_handler',
        fn: async () => {},
        then: null,
      },
    ]

    const result = toMermaid(activities)

    expect(result).toBe(`flowchart TD
    Start((start))-->
    activity1-->|then|activity2-->|catch timeout|timeout_handler
    activity2-->|then|activity3-->|catch timeout|timeout_handler
    activity3-->End((end))
    timeout_handler-->End((end))`)
  })
})

describe('toMermaidLiveEdit', () => {
  test('should generate a URL to edit a mermaid diagram', () => {
    const activities: ActivityDefinitions = [
      {
        name: 'activity1',
        start: true,
        fn: async () => {},
        then: 'activity2',
        catch: { timeout: { then: 'timeout_handler' } },
      },
      {
        name: 'activity2',
        fn: async () => {},
        then: 'activity3',
        catch: { timeout: { then: 'timeout_handler' } },
      },
      {
        name: 'activity3',
        fn: async () => {},
        then: null,
      },
      {
        name: 'timeout_handler',
        fn: async () => {},
        then: null,
      },
    ]

    const liveEditUrl = toMermaidLiveEdit(activities)

    expect(liveEditUrl).toMatchSnapshot()
  })
})

describe('toMermaidPngUrl', () => {
  test('should generate a URL to a mermaid diagram PNG', () => {
    const activities: ActivityDefinitions = [
      {
        name: 'activity1',
        start: true,
        fn: async () => {},
        then: 'activity2',
        catch: { timeout: { then: 'timeout_handler' } },
      },
      {
        name: 'activity2',
        fn: async () => {},
        then: 'activity3',
        catch: { timeout: { then: 'timeout_handler' } },
      },
      {
        name: 'activity3',
        fn: async () => {},
        then: null,
      },
      {
        name: 'timeout_handler',
        fn: async () => {},
        then: null,
      },
    ]

    const pngUrl = toMermaidPngUrl(activities)

    expect(pngUrl).toMatchSnapshot()
  })
})
