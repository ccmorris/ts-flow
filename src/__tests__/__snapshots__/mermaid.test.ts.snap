// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`toMermaid should generate a mermaid diagram 1`] = `
"flowchart TD
    Start((start))-->
    activity1[activity1]-->|then|activity2
    activity2[activity2]-->|then|activity3
    activity3[activity3]-->|then|End((end))"
`;

exports[`toMermaid should generate a mermaid diagram with catches 1`] = `
"flowchart TD
    Start((start))-->
    activity1[activity1]-->|then|activity2
    activity1-->|catch timeout|timeout_handler
    activity2[activity2]-->|then|activity3
    activity2-->|catch timeout|timeout_handler
    activity3[activity3]-->|then|End((end))
    timeout_handler[timeout_handler]-->|then|End((end))"
`;

exports[`toMermaid should generate a mermaid diagram with catches and choices 1`] = `
"flowchart TD
    Start((start))-->
    choice_task{choice task}-->|choice1|activity1
    choice_task{choice task}-->|choice2|activity2
    activity1[activity1]-->|then|activity2
    activity1-->|catch timeout|timeout_handler
    activity2[activity2]-->|then|activity3
    activity2-->|catch timeout|timeout_handler
    activity3[activity3]-->|then|End((end))
    timeout_handler[timeout_handler]-->|then|End((end))"
`;

exports[`toMermaid should style the traced tasks green 1`] = `
"flowchart TD
    Start((start)):::traced==>
    choice_task{choice task}:::traced==>|choice1|activity1
    choice_task{choice task}:::traced-->|choice2|activity2
    activity1[activity1]:::traced==>|then|activity2
    activity1-->|catch timeout|timeout_handler
    activity2[activity2]:::traced-->|then|activity3
    activity2==>|catch timeout|timeout_handler
    activity3[activity3]-->|then|End((end))
    timeout_handler[timeout_handler]:::traced==>|then|End((end)):::traced
    classDef traced fill:green,stroke-width:4px;"
`;

exports[`toMermaid should style a failed task red 1`] = `
"flowchart TD
    Start((start)):::traced==>
    choice_task{choice task}:::traced==>|choice1|activity1
    choice_task{choice task}:::traced-->|choice2|activity2
    activity1[activity1]:::traced==>|then|activity2
    activity1-->|catch timeout|timeout_handler
    activity2[activity2]:::failed-->|then|activity3
    activity2-->|catch timeout|timeout_handler
    activity3[activity3]-->|then|End((end))
    timeout_handler[timeout_handler]-->|then|End((end))
    classDef traced fill:green,stroke-width:4px;
    classDef failed fill:#cc0000,stroke-width:4px;"
`;

exports[`toMermaidLiveEdit should generate a URL to edit a mermaid diagram 1`] = `"https://mermaid.live/edit#base64:eyJjb2RlIjoiZmxvd2NoYXJ0IFREXG4gICAgU3RhcnQoKHN0YXJ0KSktLT5cbiAgICBhY3Rpdml0eTFbYWN0aXZpdHkxXS0tPnx0aGVufGFjdGl2aXR5MlxuICAgIGFjdGl2aXR5MS0tPnxjYXRjaCB0aW1lb3V0fHRpbWVvdXRfaGFuZGxlclxuICAgIGFjdGl2aXR5MlthY3Rpdml0eTJdLS0+fHRoZW58YWN0aXZpdHkzXG4gICAgYWN0aXZpdHkyLS0+fGNhdGNoIHRpbWVvdXR8dGltZW91dF9oYW5kbGVyXG4gICAgYWN0aXZpdHkzW2FjdGl2aXR5M10tLT58dGhlbnxFbmQoKGVuZCkpXG4gICAgdGltZW91dF9oYW5kbGVyW3RpbWVvdXRfaGFuZGxlcl0tLT58dGhlbnxFbmQoKGVuZCkpIiwibWVybWFpZCI6IntcInRoZW1lXCI6XCJkYXJrXCJ9IiwiYXV0b1N5bmMiOnRydWUsInJvdWdoIjpmYWxzZSwidXBkYXRlRGlhZ3JhbSI6dHJ1ZX0="`;

exports[`toMermaidPngUrl should generate a URL to a mermaid diagram PNG 1`] = `"https://mermaid.ink/img/base64:eyJjb2RlIjoiZmxvd2NoYXJ0IFREXG4gICAgU3RhcnQoKHN0YXJ0KSktLT5cbiAgICBhY3Rpdml0eTFbYWN0aXZpdHkxXS0tPnx0aGVufGFjdGl2aXR5MlxuICAgIGFjdGl2aXR5MS0tPnxjYXRjaCB0aW1lb3V0fHRpbWVvdXRfaGFuZGxlclxuICAgIGFjdGl2aXR5MlthY3Rpdml0eTJdLS0+fHRoZW58YWN0aXZpdHkzXG4gICAgYWN0aXZpdHkyLS0+fGNhdGNoIHRpbWVvdXR8dGltZW91dF9oYW5kbGVyXG4gICAgYWN0aXZpdHkzW2FjdGl2aXR5M10tLT58dGhlbnxFbmQoKGVuZCkpXG4gICAgdGltZW91dF9oYW5kbGVyW3RpbWVvdXRfaGFuZGxlcl0tLT58dGhlbnxFbmQoKGVuZCkpIiwiYXV0b1N5bmMiOnRydWUsInJvdWdoIjpmYWxzZSwidXBkYXRlRGlhZ3JhbSI6dHJ1ZX0=?type=png"`;

exports[`toMermaidPngUrl should generate a URL to a traced mermaid diagram PNG 1`] = `"https://mermaid.ink/img/base64:eyJjb2RlIjoiZmxvd2NoYXJ0IFREXG4gICAgU3RhcnQoKHN0YXJ0KSk6Ojp0cmFjZWQ9PT5cbiAgICBhY3Rpdml0eTFbYWN0aXZpdHkxXTo6OnRyYWNlZD09Pnx0aGVufGFjdGl2aXR5MlxuICAgIGFjdGl2aXR5MS0tPnxjYXRjaCB0aW1lb3V0fHRpbWVvdXRfaGFuZGxlclxuICAgIGFjdGl2aXR5MlthY3Rpdml0eTJdOjo6dHJhY2VkLS0+fHRoZW58YWN0aXZpdHkzXG4gICAgYWN0aXZpdHkyPT0+fGNhdGNoIHRpbWVvdXR8dGltZW91dF9oYW5kbGVyXG4gICAgYWN0aXZpdHkzW2FjdGl2aXR5M10tLT58dGhlbnxFbmQoKGVuZCkpXG4gICAgdGltZW91dF9oYW5kbGVyW3RpbWVvdXRfaGFuZGxlcl06Ojp0cmFjZWQ9PT58dGhlbnxFbmQoKGVuZCkpOjo6dHJhY2VkXG4gICAgY2xhc3NEZWYgdHJhY2VkIGZpbGw6Z3JlZW4sc3Ryb2tlLXdpZHRoOjRweDsiLCJhdXRvU3luYyI6dHJ1ZSwicm91Z2giOmZhbHNlLCJ1cGRhdGVEaWFncmFtIjp0cnVlfQ==?type=png"`;
